enhancements/network/multi-network-vrfs.md

```diff
@@ -0,0 +1,499 @@
+
+This functionality will continue to work, with ip rules steering the packets from the per network VRF to the appropriate egress IP VRF. CONNMARK will continue to be used so that return traffic is sent back to the correct VRF. Step 5 in the above may need to be tweaked to match on mark as well as IP, in case 2 pods have overlapping IPs, and are both egressing the same interface, with different Egress IPs. The flow would look something like this:
+
+
+![Egress IP VRF LGW](images/egress-ip-vrf-lgw.svg)
+
+
+
+#### Egress Firewall
+
+Egress firewall is enforced at the OVN logical switch, and this proposal has no effect on its functionality.
+
+
+#### Egress QoS
+
+Egress QoS is namespace scoped and functions by marking packets at the OVN logical switch, and this proposal has no effect on its functionality.
+
+
+#### Egress Service
+
+Egress service functions similar to Egress IP in local gateway mode, with the exception that all traffic paths go through the kernel networking stack. Egress Service also uses IP rules and VRFs in order to match on traffic and forward it out the right network (if specified in the CRD). It uses iptables in order to SNAT packets to the load balancer IP. Like Egress IP, with unique networks there will need to be IP rules with higher precedence to match on packets from specific networks and direct them to the right VRF.
+
+
+#### Multiple External Gateways (MEG)
+
+There will be no support for MEG or pod direct ingress on any network other than the primary default network. Remember, MEG works the same way in local or shared gateway mode, by utilizing the shared gateway path. This support may be enhanced later by extending VRFs/networks outside of the cluster.
+
+
+### Host Networked Pods
+
+By encompassing VRFs into the host, this introduces some constraints and requirements for the behavior of host networked type pods. If a host networked pod is created in a Kubernetes namespace that has a unique network, should it be restricted to only communicating with that Kubernetes network? If the answer is yes, then we would need to do the following:
+
+With Linux VRFs, the application would need to bind directly to the VRF device in order to use TCP/UDP type sockets on the network. Raw sockets by default are allowed to span multiple VRFs. This seems like a problematic solution, as we cannot expect all user applications to provide configuration for which device to bind to for TCP/UDP sockets. There is sysctl configuration that will allow TCP/UDP sockets to listen across all VRFs, but that means then a host application could communicate with any network:
```

Comment:
> Raw sockets by default are allowed to span multiple VRFs. This seems like a problematic solution

This makes it sound like you are saying that raw sockets are a problematic solution, rather than that binding directly to the VRF is problematic.

> we cannot expect all user applications to provide configuration for which device to bind to for TCP/UDP sockets

We could potentially use eBPF to intercept `bind` calls and do it for them...

> There is sysctl configuration that will allow TCP/UDP sockets to listen across all VRFs

is that likely to be acceptable to customers for the behavior of non-containerized host processes?

Reply:
> > Raw sockets by default are allowed to span multiple VRFs. This seems like a problematic solution
> 
> This makes it sound like you are saying that raw sockets are a problematic solution, rather than that binding directly to the VRF is problematic.

I'll rephrase this.

> > we cannot expect all user applications to provide configuration for which device to bind to for TCP/UDP sockets
> 
> We could potentially use eBPF to intercept `bind` calls and do it for them...
Hmm that's an interesting idea. I'll add this as potential thing we could do in the future.

> > There is sysctl configuration that will allow TCP/UDP sockets to listen across all VRFs
> 
> is that likely to be acceptable to customers for the behavior of non-containerized host processes?

I'm not sure. This seems like uncharted territory. For now I listed this as unsupported behavior because from what I've read it doesn't seem to function reliably.